net: openthread: Use Zephyr mbedTLS instead of built-in

rlubos opened the pull request:
This PR reconfigures OpenThread to use Zephyrs mbedTLS sources and configuration instead of bringing its own. This allows to easily use OpenThread with applications that already use mbedTLS for different purposes. This approach utilizes Kconfig-aware mbedTLS config file, `config-tls-generic.h`.

I've tried to bring as many optimizations, originally enabled in default OpenThread mbedTLS configuration file, as possible. This also included overriding some defaults in `Kconfig.tls-generic` specifically for OpenThread. This opens a question, whether it makes sense to have any default configuration (in terms of algorithms enabled) at all.

Originally, defaults in `Kconfig.tls-generic` were introduced to provide a seamless transition from the old, non-configurable default mbedTLS configuration file for the samples. I wonder whether we should rather drop this approach, and allow the applications to enable crypto algorithms depending on their needs, not relying on some default configuration. This, for instance, is a little bothersome when trying to use `overlay-tls.conf` along with `overlay-ot.conf`, as the former depends on defaults, while the latter disables most of them. In result, some algorithms have to be re-enabled in `overlay-tls.conf` anyway.

Resolves #3877

rlubos committed:
net: openthread: Use Zephyr's mbedTLS instead of compiling own

So far OpenThread compiled it's own, internal copy of mbedTLS library.
This commit changes this behavior by using Zephyr's mbedTLS instance
appropriately configured for OpenThread needs.

Generic mbedTLS config file was used in this case, so that application
can still configure remaining parts of mbedTLS for it's own needs.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
modules: mbedtls: Allow to enable OpenThread optimizations

OpenThread tweaks some specific mbedTLS configs for it's own need, that
allows to save some RAM/ROM. Allow to set these in generic config file
if OpenThread is enabled.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

mike-scott commented:
@rlubos thank you for pushing on this (I realize it required OpenThread side changes as well).  If I get some time this weekend, I can test it a bit.

Re: algorithm defaults.  I may have a better opinion once I take a look at how this plays with DTLS, etc.

jukkar commented on ext/lib/crypto/mbedtls/Kconfig:
```diff
@@ -58,6 +58,7 @@ endif
 
 config MBEDTLS_SSL_MAX_CONTENT_LEN
 	int "Max payload size for TLS protocol message"
+	default 768 if NET_L2_OPENTHREAD
```
One thing that is a bit controversial, is that now we introduced some networking related options to mbedtls config. Not sure how could we solve this differently.

nashif replied:
> One thing that is a bit controversial, is that now we introduced some networking related options to mbedtls config. Not sure how could we solve this differently.

agree, we need a better way to do this

rlubos replied:
@nashif @jukkar Do you have any particular solution in mind? The options we have to set this value for OpenThread needs (please correct me if I miss something):
* add `default` value, just as it's done now,
* set this value explicitly in every `overlay-ot.conf` file.

Ideally, it would be best for me if I could modify the default from the OpenThread's Kconfig file. Unfortunately, `select`/`imply` works only with boolean options to my knowledge.

jukkar replied:
Probably from these two options, adding the needed values to overlay-ot.conf file is the best. If we at some date get support for global overlays (see #14740), then we could have just one overlay-ot.conf file and managing these settings would be trivial.

nashif commented on ext/lib/crypto/mbedtls/Kconfig:
```diff
@@ -124,6 +125,7 @@ config MBEDTLS_ENABLE_HEAP
 
 config MBEDTLS_HEAP_SIZE
 	int "Heap size for mbed TLS"
+	default 8000 if NET_L2_OPENTHREAD && (OPENTHREAD_COMMISSIONER || OPENTHREAD_JOINER)
```
ditto

rlubos commented:
@mike-scott You are right, some changes were needed in OpenThread, in order to allow to use external mbedTLS heap. Fortunately, it went in quite smooth.

rlubos commented on ext/lib/crypto/mbedtls/Kconfig.tls-generic:
```diff
@@ -68,7 +68,7 @@ config MBEDTLS_KEY_EXCHANGE_RSA_PSK_ENABLED
 config MBEDTLS_KEY_EXCHANGE_RSA_ENABLED
 	bool "Enable the RSA-only based ciphersuite modes"
-	default y
+	default y if !NET_L2_OPENTHREAD
```
@nashif @jukkar Any opinion on default algorithms enabled in `Kconfig.tls-generic`? See the PR description for more details.

rlubos committed:
samples: net: Add mbedTLS tweaks in overlay-ot.conf

Add some mbedTLS tweaks for OpenThread in overlay-ot.conf.
Add sample configuration to enable Commissioner/Joiner, disabled by
default.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
Removed OT defaults from `mbedtls/Kconfig` and moved them to overlay instead.

rlubos commented:
Rebased && reworked the sample to consider the new mbedTLS location. Added DNM, until I get the SHA to update the manifest.

jukkar commented:
@rlubos should we try to get this merged to 2.1?
ping @nashif can you revisit this PR

rlubos commented:
@jukkar Yes, I'd be very happy to have this integrated. Before we can do so though, I'd need to have a dependency in mbedTLS repository merged first - https://github.com/zephyrproject-rtos/mbedtls/pull/1

jukkar commented:
ping @nashif can you re-review this

rlubos commented:
All dependencies are now merged, removing DNM.

jukkar commented:
ping @nashif can you re-check this
